File: flexinput/src/main/java/com/lytefast/flexinput/model/Attachment.java
package com.lytefast.flexinput.model;

import android.net.Uri;
import android.os.Parcel;
import android.os.Parcelable;
import android.provider.MediaStore;
import android.support.annotation.CallSuper;

import com.facebook.common.util.HashCodeUtil;


/**
 * Represents an attachable resource in the form of {@link Uri}.
 *
 * @author Sam Shih
 */
public class Attachment<T> implements Parcelable {
  protected final long id;
  protected final Uri uri;
  protected final String displayName;
  protected final T data;


  public Attachment(final long id, final Uri uri, final String displayName, T data) {
    this.id = id;
    this.uri = uri;
    this.displayName = displayName;
    this.data = data;
  }

  protected Attachment(Parcel in) {
    this.id = in.readLong();
    this.uri = in.readParcelable(Uri.class.getClassLoader());
    this.displayName = in.readString();
    this.data = null;  // this shouldn't be required anyways.
  }

  //region Getters

  public static final Creator<Attachment> CREATOR = new Creator<Attachment>() {
    @Override
    public Attachment createFromParcel(Parcel in) {
      return new Attachment(in);
    }

    @Override
    public Attachment[] newArray(int size) {
      return new Attachment[size];
    }
  };

  public long getId() {
    return id;
  }

  public Uri getUri() {
    return uri;
  }

  public String getDisplayName() {
    return displayName;
  }

  public T getData() {
    return data;
  }

  //endregion

  @Override
  public boolean equals(final Object obj) {
    if (obj instanceof Attachment) {
      Attachment other = (Attachment) obj;
      return this.id == other.id && this.uri.equals(other.uri);
    }
    return false;
  }

  @Override
  public int hashCode() {
    return HashCodeUtil.hashCode(id, uri);
  }

  @Override
  public int describeContents() {
    return 0;
  }

  @CallSuper
  @Override
  public void writeToParcel(final Parcel dest, final int flags) {
    dest.writeLong(id);
    dest.writeParcelable(uri, flags);
    dest.writeString(displayName);
//    dest.writeParcelable(data, flags);
  }
}

File: flexinput/src/main/java/com/lytefast/flexinput/model/Attachment.kt
package com.lytefast.flexinput.model

import android.content.ContentResolver
import android.net.Uri
import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.CallSuper
import androidx.core.view.inputmethod.InputContentInfoCompat
import com.facebook.common.util.HashCodeUtil
import com.lytefast.flexinput.utils.FileUtils.getFileName
import java.io.File


/**
 * Represents an attachable resource in the form of [Uri].
 *
 * @author Sam Shih
 */
open class Attachment<out T>(
    val id: Long,
    val uri: Uri,
    val displayName: String,
    val data: T? = null
) : Parcelable {

  constructor(parcelIn: Parcel) : this(
      id = parcelIn.readLong(),
      uri = parcelIn.readParcelable(Uri::class.java.classLoader) ?: Uri.EMPTY,
      displayName = parcelIn.readString() ?: "",
      data = null  // this shouldn't be required anyways.
  )

  override fun equals(other: Any?): Boolean {
    if (other != null && other is Attachment<*>) {
      return this.id == other.id && this.uri == other.uri
    }
    return false
  }

  override fun hashCode(): Int = HashCodeUtil.hashCode(id, uri)

  override fun describeContents(): Int = 0

  @CallSuper
  override fun writeToParcel(dest: Parcel, flags: Int) {
    dest.writeLong(id)
    dest.writeParcelable(uri, flags)
    dest.writeString(displayName)
    //    dest.writeParcelable(data, flags);
  }

  companion object {

    @Suppress("unused")  // Used as part of Parcellable
    @JvmField
    val CREATOR = object : Parcelable.Creator<Attachment<*>> {
      override fun createFromParcel(parcelIn: Parcel): Attachment<*> = Attachment<Any>(parcelIn)

      override fun newArray(size: Int): Array<Attachment<*>?> = arrayOfNulls(size)
    }

    @JvmStatic
    fun Uri.toAttachment(resolver: ContentResolver): Attachment<Uri> {
      val fileName = getFileName(resolver) ?: hashCode().toString()
      return Attachment(hashCode().toLong(), this, fileName, null)
    }

    @JvmStatic
    fun InputContentInfoCompat.toAttachment(
        resolver: ContentResolver,
        appendExtension: Boolean = false,
        defaultName: String = "unknown"): Attachment<InputContentInfoCompat> {
      val rawFileName = contentUri.getQueryParameter("fileName") ?: defaultName
      val fileName = rawFileName.substringAfterLast(File.separatorChar)

      val attachmentName = if (appendExtension) {
        val type = description.getMimeType(0)
            ?: contentUri.getQueryParameter("mimeType")
            ?: resolver.getType(contentUri)
        type?.let {
          "$fileName.${it.substringAfterLast('/')}"
        } ?: fileName
      } else {
        fileName
      }

      return Attachment(
          contentUri.hashCode().toLong(),
          contentUri,
          attachmentName,
          this)
    }
  }
}
